### ISSUE TITLE
NPO implementation

### ISSUE THREAD

**Issue by a-F1** [jaechan-repo/muse_bench#2]
Thank you for your excellent work.

I have some questions regarding the implementation of NPO. In the original NPO code, the implementation uses forget_loss_current - forget_loss_ref. However, in the muse code, the implementation uses forget_loss_ref - forget_loss_current. Could you clarify the reason for this difference?

ref: https://github.com/licong-lin/negative-preference-optimization/blob/main/TOFU/dataloader.py#L267

**Comment by HaomingX** [comment]
I think maybe he wrote the wrong code

**Comment by zaocan666** [comment]
I think it is because the ref code use extra CrossEntropy loss. The CrossEntropy loss contains negative sign. So the sign is correct in both implementation.

**Comment by LetheSec** [comment]
> I think it is because the ref code use extra CrossEntropy loss. The CrossEntropy loss contains negative sign. So the sign is correct in both implementation.

However,  the authors use `logits` to construct the npo loss, I think it is inaccurate.
https://github.com/jaechan-repo/muse_bench/blob/6d4fdcbdebe4ad46dccaf70f8526cd23ecff609e/baselines/baselines/iterative.py#L139

I think the author may have referenced the code in [1] (as mentioned in the [line 96](https://github.com/jaechan-repo/muse_bench/blob/6d4fdcbdebe4ad46dccaf70f8526cd23ecff609e/baselines/baselines/iterative.py#L96)). 
However, the toy model in [1] was built manually and included `softmax` in its `forward` function, but the "logits" of the huggingface model is different.

[1] https://github.com/licong-lin/negative-preference-optimization/blob/main/synthetic/mymodel.py

**Comment by HaomingX** [comment]
> > I think it is because the ref code use extra CrossEntropy loss. The CrossEntropy loss contains negative sign. So the sign is correct in both implementation.
> 
> However, the authors use `logits` to construct the npo loss, I think it is inaccurate.
> 
> https://github.com/jaechan-repo/muse_bench/blob/6d4fdcbdebe4ad46dccaf70f8526cd23ecff609e/baselines/baselines/iterative.py#L139
> 
> I think the author may have referenced the code in [1] (as mentioned in the [line 96](https://github.com/jaechan-repo/muse_bench/blob/6d4fdcbdebe4ad46dccaf70f8526cd23ecff609e/baselines/baselines/iterative.py#L96)). However, the toy model in [1] was built manually and included `softmax` in its `forward` function, but the "logits" of the huggingface model is different.
> 
> [1] https://github.com/licong-lin/negative-preference-optimization/blob/main/synthetic/mymodel.py

Hello! I wanna know why I get the shaking loss that zoom to zero with the corrected code but the smooth loss with this repo code.
The corrected code:
```
        elif self.loss_type == 'npo':
            outputs_f = model(
                x_f['input_ids'],
                labels=x_f['labels'] if 'labels' in x_f else x_f['input_ids'].clone(),
                attention_mask=x_f['attention_mask'] if 'attention_mask' in x_f else torch.ones_like(x_f['input_ids'], dtype=torch.bool)
            )
            with torch.no_grad():
                outputs_f_ref = self.ref_model(
                    x_f['input_ids'],
                    labels=x_f['labels'] if 'labels' in x_f else x_f['input_ids'].clone(),
                    attention_mask=x_f['attention_mask'] if 'attention_mask' in x_f else torch.ones_like(x_f['input_ids'], dtype=torch.bool)
                )

            outputs_f_loss = get_batch_loss(outputs_f.logits, x_f['labels'])
            outputs_f_ref_loss = get_batch_loss(outputs_f_ref.logits, x_f['labels'])
            neg_log_ratio = outputs_f_loss - outputs_f_ref_loss
            loss = -F.logsigmoid(self.beta * neg_log_ratio).mean() * 2 / self.beta
.......
def get_batch_loss(logits, labels):
    shift_logits = logits[..., :-1, :].contiguous()
    shift_labels = labels[..., 1:].contiguous()
    loss_fct = nn.CrossEntropyLoss(ignore_index=-100, reduction='none')
    loss = loss_fct(shift_logits.transpose(-1, -2), shift_labels).sum(dim=-1)
    return loss
```
The shaking loss:
![image](https://github.com/user-attachments/assets/f5ed627f-06e4-4db2-ad6b-62075553ee87)
![image](https://github.com/user-attachments/assets/7e79283f-0bcd-43d8-8365-aeb2b636c181)

but with this repo code:
![image](https://github.com/user-attachments/assets/a93399d2-2d85-407a-aa34-5f33289baa8f)

I'm confused. Really hope u can give me some advice!!!

**Comment by zaocan666** [comment]
@HaomingX It seems that there is extreme loss. Maybe you should try tuning learning rate and batch size. I have tried both ways. While this repo code results in excessive forgetting.

**Comment by HaomingX** [comment]
> @HaomingX It seems that there is extreme loss. Maybe you should try tuning learning rate and batch size. I have tried both ways. While this repo code results in excessive forgetting.

Yeah, this repo's npo behaves like GA with exceessive forgetting. I'll try more parameters. Thank u!

**Comment by LetheSec** [comment]
> > @HaomingX It seems that there is extreme loss. Maybe you should try tuning learning rate and batch size. I have tried both ways. While this repo code results in excessive forgetting.
> 
> Yeah, this repo's npo behaves like GA with exceessive forgetting. I'll try more parameters. Thank u!

Not necessarily accurate, but NPO is based on preference optimization (e.g., DPO), I think it might be more suitable for SFT models and the data in Q&A format.

**Comment by HaomingX** [comment]
> > > @HaomingX It seems that there is extreme loss. Maybe you should try tuning learning rate and batch size. I have tried both ways. While this repo code results in excessive forgetting.
> > 
> > 
> > Yeah, this repo's npo behaves like GA with exceessive forgetting. I'll try more parameters. Thank u!
> 
> Not necessarily accurate, but NPO is based on preference optimization (e.g., DPO), I think it might be more suitable for SFT models and the data in Q&A format.

Kind of makes sense, I was just unlearning with QA data. I wanna ask anothor question, should I mask the Q data when I compute batch loss for NPO. Thank u!

**Comment by LetheSec** [comment]
> Kind of makes sense, I was just unlearning with QA data. I wanna ask anothor question, should I mask the Q data when I compute batch loss for NPO. Thank u!

In the repo of NPO and TOFU, the questions are masked. Usually this is done when processing dataset, that is, replacing the token id in the question part with -100 after tokenization.
